front50-web/src/main/java/com/netflix/spinnaker/front50/controllers/PipelineController.java

```diff
@@ -237,33 +258,106 @@ public Pipeline getByApplicationAndName(
-    }
+    log.debug(
+        "Successfully validated pipeline {} in {}ms",
+        pipeline.getName(),
+        System.currentTimeMillis() - validationStartTime);
 
-    return pipelineDAO.create(pipeline.getId(), pipeline);
+    Pipeline savedPipeline = pipelineDAO.create(pipeline.getId(), pipeline);
+    log.info(
+        "Successfully saved pipeline {} in application {} in {}ms",
+        savedPipeline.getName(),
+        savedPipeline.getApplication(),
+        System.currentTimeMillis() - saveStartTime);
+    return savedPipeline;
   }
 
-  @PreAuthorize("@fiatPermissionEvaluator.isAdmin()")
+  @PreAuthorize("@fiatPermissionEvaluator.storeWholePermission()")
   @RequestMapping(value = "batchUpdate", method = RequestMethod.POST)
-  public void batchUpdate(@RequestBody List<Pipeline> pipelines) {
-    pipelineDAO.bulkImport(pipelines);
+  public Map<String, Object> batchUpdate(
+      @RequestBody List<Map<String, Object>> pipelinesJson,
+      @RequestParam(value = "staleCheck", required = false, defaultValue = "false")
+          Boolean staleCheck) {
+
+    long batchUpdateStartTime = System.currentTimeMillis();
+
+    log.debug(
+        "Deserializing the provided map of {} pipelines into a list of pipeline objects.",
+        pipelinesJson.size());
+
+    // The right side of the pair holds failed pipelines.  This needs to be
+    // List<Map<String, Object>> as opposed to List<Pipeline> since some
+    // elements of pipelineJson may fail to deserialize into Pipeline objects.
+    ImmutablePair<List<Pipeline>, List<Map<String, Object>>> deserializedPipelines =
+        deSerializePipelines(pipelinesJson);
+    List<Pipeline> pipelines = deserializedPipelines.getLeft();
+    List<Map<String, Object>> failedPipelines = deserializedPipelines.getRight();
+
+    log.info(
+        "Batch upserting the following pipelines: {}",
+        pipelines.stream().map(Pipeline::getName).collect(Collectors.toList()));
+
+    List<Pipeline> pipelinesToSave = new ArrayList<>();
+    Map<String, Object> returnData = new HashMap<>();
+
+    // List of pipelines in the provided request body which don't adhere to the schema
+    List<Pipeline> invalidPipelines = new ArrayList<>();
+    validatePipelines(pipelines, staleCheck, pipelinesToSave, invalidPipelines);
+
+    TypeReference<Map<String, Object>> mapType = new TypeReference<>() {};
+    failedPipelines.addAll(
+        invalidPipelines.stream()
+            .map((pipeline) -> objectMapper.convertValue(pipeline, mapType))
+            .collect(Collectors.toList()));
+
+    long bulkImportStartTime = System.currentTimeMillis();
+    log.debug("Bulk importing the following pipelines: {}", pipelinesToSave);
+    pipelineDAO.bulkImport(pipelinesToSave);
+    log.debug(
+        "Bulk imported {} pipelines successfully in {}ms",
+        pipelinesToSave.size(),
+        System.currentTimeMillis() - bulkImportStartTime);
+
+    List<String> savedPipelines =
```

Comment:
in the failed pipelines, you're storing more info about each pipeline, eg application, id, etc. does it make sense to add some of that information to the saved pipelines as well?

Reply:
I see where we store error information in failed pipelines, but not other stuff.  Can you provide a pointer?

Reply:
The whole failed pipeline(as a map) gets added to the list of failed pipelines in response. [Here](https://github.com/spinnaker/front50/blob/894299a067373ad289764ab0711e038ffb61ab09/front50-web/src/main/java/com/netflix/spinnaker/front50/controllers/PipelineController.java#L424-L434) and [here](https://github.com/spinnaker/front50/blob/894299a067373ad289764ab0711e038ffb61ab09/front50-web/src/main/java/com/netflix/spinnaker/front50/controllers/PipelineController.java#L329-L332) is the code for that. Additionally`errorMsg` is put into the failed pipeline map.

Reply:
@link108 - Given hundreds/thousands of pipelines, the idea I think is to let the user know more about failed pipelines and keep the info about succeeded ones to minimum. But if adding a couple of fields (like id, application) provides more value, we can do that.